### PR TITLE
Specify why on `Closed{Session,Stream}Exception`

### PR DESCRIPTION
**Pull request by trustin**
Motivation:

It is sometimes hard to know why a connection or a stream has been
closed or reset.

Modifications:

- Allow a user specify a message or a cause when creating a
  `Closed{Session,Stream}Exception`.
- Specify a message or a cause when creating `Closed{Session,Stream}Exception`
  when applicable.
- Removed redundant instantiation of `ClosedSessionException`s.
- Fixed a bug where `ClosedSessionException` is raised when
  `ClosedStreamException` is expected.

Result:

- More information when diagnosing a `Closed{Session,Stream}Exception`
- Fixed a bug where `ClosedSessionException` is raised when
  `ClosedStreamException` is expected.

### CODE REVIEW

File: core/src/main/java/com/linecorp/armeria/client/HttpSessionHandler.java [hunk 1]
```diff
@@ -322,34 +324,53 @@ public void channelInactive(ChannelHandlerContext ctx) throws Exception {
             channelPool.connect(channel.remoteAddress(), H1C, sessionPromise);
         } else {
             // Fail all pending responses.
-            final Throwable throwable;
-            if (ctx.channel().hasAttr(PENDING_EXCEPTION)) {
-                throwable = ctx.channel().attr(PENDING_EXCEPTION).get();
+            final HttpResponseDecoder responseDecoder = this.responseDecoder;
+            final Throwable pendingException;
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/client/HttpSessionHandler.java]: We can just set `pendingException` here and use it.

File: core/src/main/java/com/linecorp/armeria/client/HttpSessionHandler.java [hunk 2]
```diff
@@ -322,34 +324,53 @@ public void channelInactive(ChannelHandlerContext ctx) throws Exception {
...
+                pendingException = null;
             }
-            failUnfinishedResponses(throwable);
+
             // Cancel the timeout and reject the sessionPromise just in case the connection has been closed
             // even before the session protocol negotiation is done.
             sessionTimeoutFuture.cancel(false);
-            sessionPromise.tryFailure(throwable);
+            if (!sessionPromise.isDone()) {
+                sessionPromise.tryFailure(pendingException != null ? pendingException
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/client/HttpSessionHandler.java]: then we can just set `pendingException` here.

File: core/src/main/java/com/linecorp/armeria/client/HttpSessionHandler.java [hunk 3]
```diff
@@ -322,34 +324,53 @@ public void channelInactive(ChannelHandlerContext ctx) throws Exception {
...
+        return ClosedSessionException.get();
+    }
+
+    static void setPendingException(ChannelHandlerContext ctx, Throwable cause) {
+        final Attribute<Throwable> attr = ctx.channel().attr(PENDING_EXCEPTION);
+        final Throwable previousCause = attr.get();
+        if (previousCause == null) {
+            attr.set(cause);
+        } else if (previousCause != cause) {
+            previousCause.addSuppressed(cause);
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/client/HttpSessionHandler.java]: Is there any chance that the `previousCause` is not suppressible?

Comment (trustin) [reply core/src/main/java/com/linecorp/armeria/client/HttpSessionHandler.java]: It's possible but we need to preserve the type of the original exception, such as `UnprocessedRequestException`.

Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/client/HttpSessionHandler.java]: Yes couldn't think of that. :sweat_smile:

Comment (trustin) [reply core/src/main/java/com/linecorp/armeria/client/HttpSessionHandler.java]: Now just logging, instead of suppressing.

**Pull request merged**